src/smt.cpp

```diff
@@ -105,21 +112,199 @@ string or_omit(const expr &e) {
+  auto z3_expr = fmap(z3_ctx, [val, sz](auto ctx){ return ctx->bv_val(val, sz); });
+
+  return Expr(this, std::move(z3_expr));
+}
+
+Expr Context::bvConst(char* const name, const size_t sz) {
+  auto z3_expr = fmap(z3_ctx, [name, sz](auto ctx){ return ctx->bv_const(name, sz); });
+
+  return Expr(this, std::move(z3_expr));
+}
+
+Expr Context::boolVal(const bool val) {
+  auto z3_expr = fmap(z3_ctx, [val](auto ctx){ return ctx->bool_val(val); });
+
+  return Expr(this, std::move(z3_expr));
+}
+
+Expr::Expr(Context* const ctx, std::optional<z3::expr> &&z3_expr) : Expr(ctx) {
+  this->z3_expr = std::move(z3_expr);
+}
+
+Expr::Expr(Expr&& from) {
+  this->ctx = from.ctx;
+  this->z3_expr = std::move(from.z3_expr);
+}
+
+Expr Expr::clone() const {
+  auto cloned_z3_expr = this->z3_expr;
+  return Expr(this->ctx, std::move(cloned_z3_expr));
 }
 
 Expr Expr::simplify() const {
-  return {fmap(z3_expr, [](auto e) { return e.simplify(); })};
+  auto z3_expr = fmap(this->z3_expr, [](auto e) { return e.simplify(); });
+
+  return Expr(this->ctx, std::move(z3_expr));
 }
 
+ExprVec Expr::toNDIndices(const ExprVec &dims) const {
+  assert(dims.exprs.size() > 0);
+
+  auto idx_1d = this->clone();
+  auto expanded_exprs = ExprVec::withCapacity(this->ctx, dims.exprs.size());
+  std::for_each(dims.exprs.crbegin(), dims.exprs.crend(), 
+    [&idx_1d, &expanded_exprs](const Expr& dim) {
+      expanded_exprs.exprs.push_back(idx_1d.urem(dim));
+      idx_1d = idx_1d.udiv(dim);
+    });
+  std::reverse(expanded_exprs.exprs.begin(), expanded_exprs.exprs.end());
+  return expanded_exprs;
+}
+
+Expr Expr::urem(const Expr &rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::urem(e, *rhs.z3_expr); });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::udiv(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::udiv(e, *rhs.z3_expr); });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::add(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e + *rhs.z3_expr; });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::sub(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e - *rhs.z3_expr; });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::mul(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return e * *rhs.z3_expr; });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::ult(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::ult(e, *rhs.z3_expr); });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::ugt(const Expr& rhs) const {
+  auto z3_expr = fmap(this->z3_expr, [&rhs](auto e) { return z3::ugt(e, *rhs.z3_expr); });
+  
+  return Expr(this->ctx, std::move(z3_expr));
+}
+
+Expr Expr::boolAnd(const Expr& rhs) const {
```

Comment:
I also thought of naming it to `lAnd()`, but coming up with 'logical and' from 'lAnd' seems next to impossible :o
`logicalAnd()` or `logicAnd()` might work, but both are somewhat verbose...

Reply:
The name can simply be `and`. However, it should internally disambiguate the bitwise/logical operation!

Reply:
Hmm.. what about operator& ?

Reply:
We implemented every arithmetic operators for `Expr` as methods, so I'm kinda concerned that it may break the consistency...

Reply:
operator& can be also declared as a member function, so it is fine.